Feature/login view

LehlohonoloIsaac opened the pull request:
<img width="369" alt="screen shot 2018-05-14 at 15 48 13" src="https://user-images.githubusercontent.com/13454525/40001374-611783c0-578e-11e8-94c3-c3785cc187e9.png">

LehlohonoloIsaac committed:
Create a dependency container

--> Update classes and protocols to use property injection.

LehlohonoloIsaac committed:
Create Login UI

LehlohonoloIsaac committed:
Merge branch 'feature/login-UI' into feature/login-view

LehlohonoloIsaac committed:
Update class name LoginVC to LoginViewController

LehlohonoloIsaac committed:
Create an alert when there's an error when the user tries to log in

LehlohonoloIsaac committed:
Merge remote-tracking branch 'origin/develop' into feature/login-view

LehlohonoloIsaac committed:
Update the run script

pjwelcome commented:
@VaalSquared the build is failing

pjwelcome commented on Showcase-iOS/AppDelegate.swift:
```diff
@@ -16,7 +16,11 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
     func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplicationLaunchOptionsKey: Any]?) -> Bool {
         FirebaseApp.configure()
-        // Override point for customization after application launch.
+        let loginView = LoginRouter().createModule()
```
Why does this need to be done, why can it not just pick up the root view controller from the storyboard? I dont think this is needed. Also setting the frame of the window is bad practice. everything should be done with constraints as this could sometimes fail. But I think this should be removed.

pjwelcome commented on Showcase-iOS/Container/DependencyContainer.swift:
```diff
@@ -0,0 +1,32 @@
+import FirebaseAuth
+import Swinject
+
+class DependencyContainer {
```
Please can you add the container in a separate pull request as it has nothing to do with LoginView feature. This is what I asked you about before making the PR, if you were going to make 2 PR's one for the container and another for the login feature and now you have bundled it together.

pjwelcome commented on Showcase-iOS/Login/Interactor/LoginInteractor.swift:
```diff
@@ -9,20 +9,15 @@
-        self.loginPresenter = loginPresenter
-        self.userAuthenticator = userAuthenticator
-    }
+    var loginPresenter: LoginInteractorPresentable?
```
I think this whole property inject is not the best approach, you have now introduced optional types for all interfaces. If for some reason it doesnt get set, you will land up with a crash.  This also forces nil checks when you shouldnt need to check as the class shouldnt be created without its dependencies.

pjwelcome commented on Showcase-iOS/Login/Router/LoginRouter.swift:
```diff
@@ -7,10 +7,14 @@
-        return LoginVC.instantiate(fromAppStoryboard: .Login)
+        let loginView = DependencyContainer()
+                        .container?
+                        .resolve(LoginPresenterViewable.self) as! LoginViewController
```
unwrapping.

pjwelcome commented on Showcase-iOS/Login/View/LoginViewController.swift:
```diff
@@ -0,0 +1,143 @@
+    func showEmailValidationFailure(withError error: AuthenticationError) {
+        alertController?.title = "Invalid email."
+        alertController?.message = "Please provide a valid email."
+        present(alertController!, animated: true, completion: nil)
```
unwrapping

pjwelcome commented on Showcase-iOS/Login/View/LoginViewController.swift:
```diff
@@ -0,0 +1,143 @@
+    }
+    
+    func showPasswordValidationFailure(withError error: AuthenticationError) {
+        alertController?.title = "Invalid password."
```
This code looks like its duplicated in 3 places, please put it in a function that takes a message and title

pjwelcome commented on Showcase-iOS/Login/View/LoginViewController.swift:
```diff
@@ -0,0 +1,143 @@
+    }
+    
+    func showSuccess() {
+        
```
remove empty methods

pjwelcome commented on Showcase-iOSTests/Login/New Group/LoginVCTests.swift:
```diff
@@ -0,0 +1,91 @@
+    
+    func testThatWhenViewWillDisappearMethodGetsCalledThenRemoveObserversIsInvoked() {
+        stub(mockKeyBoardObservable) { (mock) in
+            _ = when(mock.removeObservers().then {})
```
why can you not use "thenDoNothing" here like you did for the others

LehlohonoloIsaac committed:
Update PR

--> Remove dependency container from PR.
--> Present the Login View directly from storyboard.
--> Create a function to display an alert in LoginViewController.
--> Update the showSuccess to perform segue

LehlohonoloIsaac committed:
Update tests for LoginViewController

pjwelcome commented on Showcase-iOSTests/Login/New Group/LoginVCTests.swift:
```diff
@@ -0,0 +1,85 @@
+import Cuckoo
+@testable import Showcase_iOS
+
+class LoginVCTests: XCTestCase {
```
the name of this class. Please can you name it properly. We don't do abbreviations. Please name it LoginViewControllerTests. Naming is important can we please name classes correctly.

LehlohonoloIsaac committed:
Update the name of LoginVCTests to LoginViewControllerTests
